lib/tasks/cards.rake

```diff
@@ -222,45 +221,303 @@ namespace :cards do
+        raise ActiveRecord::Rollback
+      end
+
+      num_assoc = 0
+      card_pool_id_to_card_id.each_slice(1000) { |m|
+        num_assoc += m.length
+        puts '  %d card_pool -> card associations' % num_assoc
+        sql = "INSERT INTO card_pools_cards (card_pool_id, card_id) VALUES "
+        vals = []
+        m.each { |m|
+          # TODO(ams): use the associations object for this or ensure this is safe
+          vals << "('%s', '%s')" % [m[0], m[1]]
+        }
+        sql << vals.join(", ")
+        unless ActiveRecord::Base.connection.execute(sql)
+          puts 'Hit an error while inserting card_pool -> card mappings. Rolling back.'
+          raise ActiveRecord::Rollback
+        end
+      }
+    end
+  end
+
+  def import_restrictions(restrictions)
+    new_restrictions = []
+    restrictions.each { |m|
+      new_restrictions << Restriction.new(
+        id: m['code'],
+        name: m['name'],
+        date_start: m['date_start'],
+        point_limit: m['point_limit']
+      )
+    }
+    Restriction.import new_restrictions, on_duplicate_key_update: { conflict_target: [ :id ], columns: :all }
+  end
+
+  def import_restriction_cards(restrictions)
+    # restriction_cards = []
+    # restrictions.each { |m|
+    #   next if m['cards'].nil?
+    #   m['cards'].each do |code, card|
+    #     restriction_cards << RestrictionCard.new(
+    #       id: m['code'] + '_' + code,
+    #       restriction_id: m['code'],
+    #       card_id: code,
+    #       global_penalty: card["global_penalty"] || 0,
+    #       universal_faction_cost: card["universal_faction_cost"] || 0,
+    #       is_restricted: card["is_restricted"] || false,
+    #       is_banned: card["deck_limit"] == 0,
+    #       points: card["points"] || 0
+    #     )
+    #   end
+    # }
+    # RestrictionCard.import restriction_cards, on_duplicate_key_update: { conflict_target: [ :id ], columns: :all }
+  end
+
+  def import_restriction_subtypes(restrictions)
+    # restriction_subtypes = []
+    # restrictions.each { |m|
+    #   next if m['subtypes'].nil?
+    #   m['subtypes'].each do |code, subtype|
+    #     restriction_subtypes << RestrictionSubtype.new(
+    #       id: m['code'] + '_' + code,
+    #       restriction_id: m['code'],
+    #       subtype_id: code,
+    #       global_penalty: subtype["global_penalty"] || 0,
+    #       universal_faction_cost: subtype["universal_faction_cost"] || 0,
+    #       is_restricted: subtype["is_restricted"] || false,
+    #       is_banned: subtype["deck_limit"] == 0,
+    #       points: subtype["points"] || 0
+    #     )
+    #   end
+    # }
+    # RestrictionSubtype.import restriction_subtypes, on_duplicate_key_update: { conflict_target: [ :id ], columns: :all }
+  end
+
+  def import_snapshots(formats)
+    snapshots = []
+    formats.each { |f|
+      f['snapshots'].each_with_index do |s, i|
+        snapshots << Snapshot.new(
+          id: generate_snapshot_code(f, i),
```

Comment:
this leaves things a bit fragile if the order of the JSON data changes.

Reply:
I can make this sort the snapshots by date first, or I can explicitly add a `code` field to each snapshot. What do you think would be best?

Reply:
the thing that matches the convention for everything else is to have an id field in the JSON and use that.

I don't think a numeric ID is bad here, but making the JSON the source of truth for the id seems like the way to go.

Reply:
I'll make it an `id` field to match your PR.

Reply:
Should the IDs be `standard_0`, `standard_1`, etc. or `standard_000`, `standard_001`, etc. to account for alphanumeric ordering?

Reply:
they can be whatever.  The ordering could be done by date regardless of what the id is.

Perhaps you could generate the id as [date.gsub('-', '_'), card_pool, restriction].join("_") ? if we don't need to expose the id, but just use it, that gives us something consistent without as many changes?

It's up to you.

Please do something similar for the format snapshot ids as well.